# Ecommerce Purchases Exercise Overview

Hi Guys,<br>
After a crash course on pandas for data analysis, its time to do some practice!!!<br>
Because of privacy issues, I have created a fake dataset here with 30K entries. <br>
The situation is, customers are providing some personal information while purchasing stuff on-line or in-store. For some reasons, your client wants to know the answer to some of his questions from the dataset, let's try to help him!.<br>

*Feel free to consult the solutions if needed. Please note, the tasks given in the exercises, can be solved in different ways. Try your best answer and compare with the solutions.* <br>
&#9758; The dataset is provided in the course downloads. 

**1. Import Pandas and Read the csv file.**

In [1]:
import pandas as pd
import numpy as np
file_path = 'Cust_Purch_FakeData.csv'
df = pd.read_csv(file_path)
df

prefix    first       last              email  gender  age  \
0        Dr.      Ray     Morton    sebvajom@kol.km    Male   38   
1       Miss  Claudia  Rodriquez      acu@jatsot.ug  Female   51   
2       Miss    Harry      Meyer          zuz@lo.wf  Female   51   
3       Miss    Edith    Gilbert  hansohsi@jupec.md  Female   55   
4        Dr.     Lura     Murphy     webediti@je.st  Female   20   
...      ...      ...        ...                ...     ...  ...   
29995    Mr.     John    Maxwell  larlisu@sejenu.ad  Female   38   
29996    Dr.   Alfred       Bass      le@duridzo.gh  Female   24   
29997   Mrs.   Edward     Curtis      ta@kucemah.et  Female   56   
29998    Mr.      Max      Woods       li@mincoz.mp  Female   64   
29999    Mr.     Olga  Rodriguez         ner@ota.ec    Male   22   

                             company                profession  \
0                     Medtronic Inc.          Health Therapist   
1       Ames Department Stores, Inc.          Health Therapist   
2                          CSX Corp.       Political Scientist   
3             Murphy Oil Corporation    Transportation Manager   
4                       PETsMART Inc              Statistician   
...                              ...                       ...   
29995                  Spherion Corp       Publishing Director   
29996                  Ford Motor Co          Diamond Merchant   
29997  Park Place Entertainment Corp        Executive Director   
29998           Nacco Industries Inc  University Administrator   
29999          Texas Industries Inc.           Leasing Manager   

                phone   postal province             cc_no   cc_exp  \
0      (987) 619-2695  B6V 3W3       MB  5020000000000230  05/2018   
1      (356) 736-7352  G7M 5F3       SK  5020000000000230  07/2028   
2      (539) 246-1806  A0Z 6P9       NS  6300000000000000  02/2023   
3      (984) 962-7494  P9I 9H3       YT  3530000000000000  02/2028   
4      (902) 568-9748  S1A 6K0       ON  4030000000000000  10/2025   
...               ...      ...      ...               ...      ...   
29995  (873) 298-1178  V4U 0X2       YT  6330000000000000  04/2028   
29996  (541) 904-3323  T9O 1H6       PE   201000000000000  12/2026   
29997  (375) 931-8605  X5Z 4E1       QC   201000000000000  06/2021   
29998  (786) 481-9627  G0S 9Q8       NS    30000000000000  08/2023   
29999  (841) 962-4087  N2D 5K3       BC  6300000000000000  05/2024   

                         cc_type  price(CAD) fav_color              ip  \
0                           Solo        8.36      Blue    126.23.139.2   
1                           Visa       68.31     Black  106.198.76.211   
2                         Switch       34.65     Black  186.150.187.29   
3                        Maestro       64.59     White   80.140.57.161   
4      Diners Club International       20.83    Yellow   211.103.43.41   
...                          ...         ...       ...             ...   
29995              Visa Electron       90.37       Red  194.14.119.109   
29996           American Express        5.94     White  154.114.74.254   
29997                    Maestro       65.66     Black   188.0.106.220   
29998                       Solo       52.66     Black  129.168.247.27   
29999               InstaPayment       17.71    Yellow     70.2.107.65   

         weekday ampm        date  
0         Sunday   pm  04/05/1930  
1        Tuesday   am  12/20/1926  
2      Wednesday   pm  08/20/1931  
3       Saturday   am  06/18/2001  
4         Friday   pm  06/14/2045  
...          ...  ...         ...  
29995     Sunday   am  04/18/2014  
29996     Sunday   pm  09/27/2068  
29997     Monday   pm  09/07/1997  
29998     Monday   pm  09/08/1938  
29999  Wednesday   am  06/22/1956  

[30000 rows x 20 columns]

**2. Its good idea to see how the data look like, display first 5 rows of your data-set.**

In [2]:
# Code here please
df.head()

prefix    first       last              email  gender  age  \
0    Dr.      Ray     Morton    sebvajom@kol.km    Male   38   
1   Miss  Claudia  Rodriquez      acu@jatsot.ug  Female   51   
2   Miss    Harry      Meyer          zuz@lo.wf  Female   51   
3   Miss    Edith    Gilbert  hansohsi@jupec.md  Female   55   
4    Dr.     Lura     Murphy     webediti@je.st  Female   20   

                        company              profession           phone  \
0                Medtronic Inc.        Health Therapist  (987) 619-2695   
1  Ames Department Stores, Inc.        Health Therapist  (356) 736-7352   
2                     CSX Corp.     Political Scientist  (539) 246-1806   
3        Murphy Oil Corporation  Transportation Manager  (984) 962-7494   
4                  PETsMART Inc            Statistician  (902) 568-9748   

    postal province             cc_no   cc_exp                    cc_type  \
0  B6V 3W3       MB  5020000000000230  05/2018                       Solo   
1  G7M 5F3       SK  5020000000000230  07/2028                       Visa   
2  A0Z 6P9       NS  6300000000000000  02/2023                     Switch   
3  P9I 9H3       YT  3530000000000000  02/2028                    Maestro   
4  S1A 6K0       ON  4030000000000000  10/2025  Diners Club International   

   price(CAD) fav_color              ip    weekday ampm        date  
0        8.36      Blue    126.23.139.2     Sunday   pm  04/05/1930  
1       68.31     Black  106.198.76.211    Tuesday   am  12/20/1926  
2       34.65     Black  186.150.187.29  Wednesday   pm  08/20/1931  
3       64.59     White   80.140.57.161   Saturday   am  06/18/2001  
4       20.83    Yellow   211.103.43.41     Friday   pm  06/14/2045

**3. How many entries your data have? 
Can you tell the no. of columns in your data?**

In [3]:
# Code here please
df_info = df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 20 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   prefix      30000 non-null  object 
 1   first       30000 non-null  object 
 2   last        30000 non-null  object 
 3   email       30000 non-null  object 
 4   gender      30000 non-null  object 
 5   age         30000 non-null  int64  
 6   company     30000 non-null  object 
 7   profession  30000 non-null  object 
 8   phone       30000 non-null  object 
 9   postal      30000 non-null  object 
 10  province    30000 non-null  object 
 11  cc_no       30000 non-null  int64  
 12  cc_exp      30000 non-null  object 
 13  cc_type     30000 non-null  object 
 14  price(CAD)  30000 non-null  float64
 15  fav_color   30000 non-null  object 
 16  ip          30000 non-null  object 
 17  weekday     30000 non-null  object 
 18  ampm        30000 non-null  object 
 19  date        30000 non-nul

**4. What are the max and min ages of your customer? Can you find mean of your customer?**

In [4]:
# Code here please
max_age = df['age'].max()
min_age = df['age'].min()
mean_age = df['age'].mean()
print(f"Max. age of the customer is: {max_age}")
print(f"Min. age of the customer is: {min_age}")
print(f"Avg. age of the customer is: {mean_age}")

Max. age of the customer is: 65
Min. age of the customer is: 18
Avg. age of the customer is: 41.550066666666666


**5. What are the three most common customer's names?**<br>
&#9989; [<code>**value_counts()**</code>](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) returns object containing counts of unique values. The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default.

In [5]:
# Code here please


**6. Two customers have the same phone number, can you find those customers**?

In [6]:
# Code here please

In [7]:
# lets find what phone number is twice!


In [8]:
# Code here please

In [9]:
# Now we know the phone number, let's find out the other stuff!


**7. How many customers have profession "Structural Engineer"?**

In [10]:
# Code here please

**8. How many male customers are 'Structural Engineer'?**

In [11]:
# Code here please

**9. Find out the female Structural Engineers from province Alberta (AB)?** 

In [12]:
# Code here please

**10. What is the max, min and average spending?**

In [13]:
# Code here please

**11. Who did not spend anything? Company wants to send a deal to encourage the customer to buy stuff!**

In [14]:
# Code here please

**12. As a loyalty reward, company wants to send thanks coupon to those who spent 100CAD or more, please find out the customers?**

In [15]:
# Code here please

**13. How many emails are associated with this credit card number '5020000000000230'?**

In [16]:
# Code here please

**14. We need to send new cards to the customers well before the expire, how many cards are expiring in 2019?**<br>
*Use `sum()` and `count()` and see the difference in their use :)*

In [17]:
# Code here please

In [18]:
#cust[cust['cc_exp'].apply(lambda x: x[5:]) == '19'].count()


**15. How many people use Visa as their Credit Card Provider?**

In [19]:
# Code here please

**16. Can you find the customer who spent 100 CAD using Visa?**

In [20]:
# Code here please

**17. What are two most common professions?**

In [21]:
# Code here please

**18. Can you tell the top 5 most popular email providers? (e.g. gmail.com, yahoo.com, etc...)**

In [22]:
# Code here please

**19. Is there any customer who is using email with "am.edu"?**<br>
Hint: Use `lambda` expression in `apply()`. split the email address at `@`.  

In [23]:
# Code here please

**20. Which day of the week, the store gets more customers?**

In [24]:
# Code here please

# Excellent work!